transformer_engine/pytorch/csrc/extensions.cu

```diff
@@ -1767,6 +1767,174 @@ bool userbuf_comm_available() {  // TODO(ksivamani) check on python side
 
 void placeholder() {}  // TODO(ksivamani) clean this up
 
+namespace flash_attention {
+
+constexpr int warp_size = 32;
+constexpr int type_size = 2;  // FP16 or BF16
+constexpr int nvec = sizeof(uint64_t) / type_size;
+constexpr int load_size = warp_size * nvec;
+constexpr int block_size = 512;
+
+template <typename T>
+__launch_bounds__(block_size)
+__global__ void prepare_kernel(const T *qkvi,
+                               T *qkv,
+                               const size_t B,
+                               const size_t S,
+                               const size_t Z,
+                               const size_t W) {
+    const int warpid = (blockDim.x * blockIdx.x + threadIdx.x) / warp_size;
+    const int id_in_warp = threadIdx.x % warp_size;
+    const size_t offset_input = blockIdx.y * W + warpid * 3 * W * Z + id_in_warp * nvec;
+    const T *my_input = qkvi + offset_input;
+
+    const size_t s = warpid / B;
+    if (s >= S) return;
+
+    const size_t b = warpid % B;
+
+    const size_t offset_output = blockIdx.y * B * S * Z * W +
+                                 (s + b * S) * W * Z +
+                                 id_in_warp * nvec;
+
+    T *my_output = qkv + offset_output;
+
+    for (int i = 0; i < Z; ++i) {
+        uint64_t *out = reinterpret_cast<uint64_t*>(my_output + i * load_size);
+        *out = *reinterpret_cast<const uint64_t*>(my_input + i * load_size * 3);
+    }
+}
+
+template <typename T>
+__launch_bounds__(block_size)
+__global__ void prepare_kernel_bwd(const T *q, const T *k, const T *v,
+                                   T *qkv, const size_t B, const size_t S,
+                                   const size_t Z, const size_t W) {
+    const T *input = blockIdx.y == 0 ? q : (blockIdx.y == 1 ? k : v);
+
+    const int warpid = (blockDim.x * blockIdx.x + threadIdx.x) / warp_size;
+    const int id_in_warp = threadIdx.x % warp_size;
+    const size_t offset_input = warpid * W * Z + id_in_warp * nvec;
+    const T *my_input = input + offset_input;
+
+    const size_t b = warpid / S;
+    if (b >= B) return;
+
+    const size_t s = warpid % S;
+
+    const size_t offset_output = (b + s * B) * 3 * W * Z +
+                                 id_in_warp * nvec + blockIdx.y * W;
+
+    T *my_output = qkv + offset_output;
+
+    for (int i = 0; i < Z; ++i) {
+        uint64_t *out = reinterpret_cast<uint64_t*>(my_output + i * load_size * 3);
+        *out = *reinterpret_cast<const uint64_t*>(my_input + i * load_size);
+    }
+}
+
+}  // namespace flash_attention
+
+at::Tensor fa_prepare(at::Tensor qkvi) {
+    NVTE_CHECK(qkvi.dim() == 4, "Expected 4-dim tensor.");
+    NVTE_CHECK(qkvi.scalar_type() == at::ScalarType::Half ||
+               qkvi.scalar_type() == at::ScalarType::BFloat16);
+    NVTE_CHECK(qkvi.size(3) % flash_attention::load_size == 0);
```

Comment:
This check is redundant given we check for equality in next line

Reply:
Yes, but the next step in the following PR is going to be generalizing it to all sizes being 0 modulo this, so I left this check here and as part of that generalization will remove the next check.